[28기 정해수] 로그인 페이지 유효성 검사 및 메인페이지 댓글 추가 기능 구현 완료.

wjdgotn77 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [x] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 로그인 페이지
  - 아이디, 비밀번호 입력창을 사용한 로그인 기능 구현
  - ID 및 PW 유효성 검사
- 메인 페이지
  - 댓글 추가 기능 구현
  - Mock Data 사용 기존 댓글 추가 

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 로그인 입력 데이터 유효성 검사 
- 입력창에 변화가 일어날 때 마다 유효성 검사 함수가 실행된다.
- 유효성 검사 함수의 반환값(boolean)에 따라 버튼의 배경색과 disabled 속성이 변화한다.
- 테스트 방법 : `/login-haesoo` 페이지 이동 >>> 로그인 input 데이터 입력

2. 댓글 추가 기능 및 Mock Data 를 활용해 기존 댓글 화면에 구현.
- 댓글을 작성하고 엔터 또는 게시 버튼을 누르면 댓글이 추가된다.
- Mock Data 내에 작성되어진 기존 댓글들이 화면에 출력.
- 테스트 방법 : '/Main-haesoo' 페이지 이동 >>> 댓글 입력

3. 스토리 페이지 Mock Data를 활용.
- 피드가 화면에 그려질 때, storyData.json 에서 Mock Data를 받아와 StoryFeed 컴포넌트로 UI를 그려준다. 

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
### 고민 및 어려웠던 점

- useEffect를 어디서 사용해서 Mock Data를 state 로 관리해주고 해당 state를 AddComment의 props 로 받아와 사용해주어야하는지에 대해 고민했습니다.
- 처음 vanilla JS로 구현시 기존에 작성되어져 보이는 댓글과 새로 입력했을 때 입력되는 댓글의 `<div >`가 나눠 작성되어 있어 컴포넌트로 만들어 작업해주는데 있어서 구조를 잡아보는 것이 낯설었지만 , 큰 `Comment` 컴포넌트를 만들고 내부에 댓글을 추가할 수 있는 `AddComment`를 작성해주었습니다
- Style 관련 클래스 네이밍을 통일되게끔 작성해보려했으나 어려웠습니다.

### 새로 알게된 점

- input 태그의 name, required 속성에 대해 알게되었고 사용하였습니다.
- Data의 URL 주소와 같은 string으로 사용되는 부분은 변수에 저장해주고 해당 변수를 사용해주면 유지보수 측면에서 도움이 됩니다.
- Id와 pw를 하나의 state로 관리하는데에 있어 구조분해할당 내부에 계산된 속성명을 사용할 수 있는 것을 알게되었습니다.

<br />

## :: 기타 질문 및 특이 사항

- SASS를 사용하면서 훨씬 줄어들었지만 현재 작성된 클래스명이 적절한지 고민됩니다.
- 현재 Feed 부분을 컴포넌트로 만들어주고 있는데, 가장 큰 `Main`에서 useEffet로 하나의 Mock Data를 받아와 state 로 관리하고 props로 전달해주려고 하는데, 범위가 커지니 조금 어렵게 느껴집니다.

## :: 코드 리뷰 반영

- 불필요한 div 태그들을 삭제하였습니다.
- 컴포넌트 별 클래스명을 다시 작성하였습니다.
-  state 관련 변수명을 의미가 명확할 수 있도록 수정하였습니다.
- 반복되는 부분을 컴포넌트로 묶고 map을 사용했습니다.
- Login 컴포넌트의 disabled 속성에 truthy/ falsy 한 값을 가지는 변수를 넣어, 코드를 정리하였습니다.
- 태그 등의 정보를 담고 있는 객체의 변수명을 상수로 바꾸고 컴포넌트 생성 함수 밖으로 빼, 함수가 실행될 때마다 객체를 불러질 수 있는 상황을 수정하였습니다.
- p태그의 width 값을 100%로 주었던 것을 지웠습니다.
- Login style에서 요소 자체의 width를 100%로 주고 불필요한 flex는 삭제하였습니다.

wjdgotn77 committed:
[Add: Login, Main의 js 와 scss 추가]

wjdgotn77 committed:
[Fix: reset.scss 충돌 수정]

wjdgotn77 committed:
[Style: Login, Main 의 scss import]

wjdgotn77 committed:
[Style: Login Footer컴포넌트 position 속성 추가]

wjdgotn77 committed:
[Add: Footer컴포넌트 추가]

wjdgotn77 committed:
[Add: Mission 1 - 사용자 입력 데이터 저장 기능 구현]

wjdgotn77 committed:
[Add: Mission 2 - 로그인 버튼 활성화 기능 구현]

wjdgotn77 committed:
[Add : Mission 4 - 댓글 컴포넌트화, props로 데이터 전달 구현]

wjdgotn77 committed:
[Add : Mission 4 - 댓글 컴포넌트화, props로 데이터 전달 구현 + Key추가]

wjdgotn77 committed:
[Modify: a tag 수정]

wjdgotn77 committed:
[Add: 컴포넌트 분리 및 목데이터 사용과 css 수정]

- style
 Login.scss 태그 수정.
 Aside, MainLeft,Feed, Comment, AddComment, StoryFeed 컴포넌트 별 스타일 분리 작성.

- js
 Comment : useState 및 useEffect 사용하여 목데이터 및 댓글 추가 기능 구현.
 Main, Aside, MainLeft, Feed, Comment, AddComment, StoryFeed 컴포넌트 분리작성.

wjdgotn77 committed:
[Modify: Login- js / scss 리팩토링 체크리스트 참고 수정]

wjdgotn77 committed:
[Modify: Main.scss 리팩토링 체크리스트 참고 수정]

wjdgotn77 committed:
[Modify: Main JS/Style  리팩토링 체크리스트 참고 수정]

wjdgotn77 committed:
[Modify: Main JS 주석 삭제]

wjdgotn77 committed:
[Modify: AddComment 구조분해할당 사용]

wjdgotn77 committed:
[Modify: esLint self closing 수정]

wjdgotn77 committed:
[Modify: 구조분해 할당 사용]

Runor0624 commented on src/pages/haesoo/Login/Login.js:
```diff
@@ -0,0 +1,76 @@
+    navigate('/main-haesoo');
+  };
+
+  const handleIdChange = event => {
+    setIsIdValid(event.target.value);
+  };
+
+  const handlePwChange = event => {
+    setIsPwValid(event.target.value);
+  };
```
체크리스트 4번 참고해서 수정 부탁드립니다.

Runor0624 commented on src/pages/haesoo/Main/Aside/Aside.js:
```diff
@@ -0,0 +1,108 @@
+          </div>
+        </div>
+        <footer className="main-footer">
+          <div className="footer-tags">
+            <span>소개 ・</span>
+            <span>도움말 ・</span>
+            <span>홍보 ・</span>
+            <span>개인정보처리방침 ・</span>
+            <span>약관 ・</span>
+            <span>위치 ・</span>
+            <span>인기 ・</span>
+            <span>계정 ・</span>
+            <span>해시태그</span>
```
체크리스트 6번과 유사한 사례같습니다
체크리스트 참고 부탁드립니다.

Runor0624 reviewed (commented):
리뷰사항 2개 참고 부탁드립니다.

minukk commented on src/pages/haesoo/Main/MainLeft/StoryFeed/StoryFeed.js:
```diff
@@ -0,0 +1,51 @@
+        <div className="story__id">bona_is_cute</div>
+      </div>
+      <div className="story-container__contents">
+        <div className="story__image">
```
해수님! 중복되는 story를 Array.map()을 이용해 반복하면 어떨까요?

minukk commented on src/pages/haesoo/Login/Login.js:
```diff
@@ -0,0 +1,76 @@
+                type="password"
+              />
+              <button
+                disabled={checkedValid(isIdValid, isPwValid)}
```
checkedValid에서 인자 두개로 전달해주는 방법 말고 checkedValid 함수 안에서 isIdValid와 isPwValid를 쓰는 방법은 어떤가요?

minukk reviewed (commented):
westagram 프로젝트 수고많으셨습니다. ㅎㅎ

YunKukPark commented on src/pages/haesoo/Login/Login.js:
```diff
@@ -0,0 +1,76 @@
+    navigate('/main-haesoo');
+  };
+
+  const handleIdChange = event => {
```
Input 을 두개 같이 핸들링 하고 있는데 같은 코드가 반복되는 느낌이군요!! 🤔
혹시 이렇게 하는건 어떠세요??
```jsx
 const handleInput = event => {
    const {
      target: { name, value },
    } = event;
    if (name === 'user-id') setId(value);
    if (name === 'user-password') setPassword(value);

    handleButton(id, password);
  };
```
`handleInput` 은 Input을 관리하는 녀석이고 input에 부여한 name 이나 id 값에 따라 handling 을 따로 해주는 거죠!

YunKukPark commented on src/pages/haesoo/Login/Login.js:
```diff
@@ -0,0 +1,76 @@
+    setIsPwValid(event.target.value);
+  };
+
+  const checkedValid = (id, pw) => {
+    return id.includes('@') && pw.length > 6 ? false : true;
+  };
```
한 함수에서 두가지의 validation을 해주는 것 보다
```jsx
const checkId = (id) => id.includes('@') ? true : false;
const checkPw = (pw) => pw.lenght > 6 ? ture : false;
const ValidateLogin = () => {
  const isValidate = checkId(id) && checkPw(pw)
  return isValidate;
}
```
이런 식으로 따로 분리해서 처리해주는건 어떤가요??
1함수 1동작 !

YunKukPark commented on src/pages/haesoo/Main/MainLeft/Feed/Comment/Comment.js:
```diff
@@ -0,0 +1,59 @@
+    setCommentInput(value);
+  };
+
+  const handleSubmit = event => {
+    event.preventDefault();
+    setComment(
+      comments.concat({
+        id: '',
+        userName: 'haesoo',
+        content: commentInput,
+        isLiked: '',
+      })
+    );
+    setCommentInput('');
+  };
```
> 가독성 향상을 위해! 

concat 안에 배열의 줄 수가 긴데, 한줄로 줄여보는건 어떠신가요??
```suggestion
 const newComment = {
   id: '',
   어쩌구저쩌구
   ...
 }
 
 const handleSubmit = event =>{
   event.preventDefault(); 
   setComment(newComment, [...comments]);
   setCommentInput(''); 
 }
```

YunKukPark commented on src/pages/haesoo/Main/MainLeft/Feed/Feed.scss:
```diff
@@ -0,0 +1,169 @@
+    width: 100%;
+    height: 780px;
+
+    .image__upload {
+      width: 100%;
+      height: 760px;
+    }
+
+    .image__upload {
+      img {
+        width: 100%;
+        height: 100%;
+      }
+    }
```
```suggestion
    .image__upload {
      width: 100%;
      height: 760px;
      
        img {
        width: 100%;
        height: 100%;
        }
    }

   
```

YunKukPark commented on src/pages/haesoo/Main/MainLeft/Feed/Feed.scss:
```diff
@@ -0,0 +1,169 @@
+    }
+  }
+
+  .feed-icons {
+    display: flex;
+    justify-content: center;
+    align-items: center;
+    width: 100%;
+    padding: 10px 0px;
+
+    .feed-icons__left {
+      width: 50%;
+      left: 0px;
+      padding: 0px 20px;
+      cursor: pointer;
+
+      img {
+        width: 20px;
+        margin-right: 10px;
+      }
+    }
+
+    .feed-icons__right {
+      width: 50%;
+      right: 0;
+      padding: 0px 10px;
+      text-align: right;
+
+      img {
+        width: 20px;
+        margin-right: 10px;
+      }
+    }
+  }
```
sass 사용 시 `&` 연신자를 사용해서 같은 글자는 줄이도록 해요! 

```suggestion
  .feed-icons {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
    padding: 10px 0px;

    &__left {
      width: 50%;
      left: 0px;
      padding: 0px 20px;
      cursor: pointer;

      img {
        width: 20px;
        margin-right: 10px;
      }
    }

    &__right {
      width: 50%;
      right: 0;
      padding: 0px 10px;
      text-align: right;

      img {
        width: 20px;
        margin-right: 10px;
      }
    }
  }
```

wjdgotn77 committed:
[Refactor: 피어리뷰 후 Feed scss &연산자 적용하여 수정]

wjdgotn77 committed:
[Refactor: 피어리뷰 후 map 사용 반영]

wjdgotn77 committed:
[Refactor: 피어리뷰 후 리팩토링]

wjdgotn77 committed:
[Refactor: 피어리뷰 후 수정사항 반영 / map 사용]

wjdgotn77 committed:
[Modify: key값 추가]

wjdgotn77 committed:
[Add: 로그인 및 회원가입 실습]

wjdgotn77 committed:
[Modify: Feed.js 컴포넌트 분리중]

wjdgotn77 committed:
[Modify: Feed컴포넌트 분리 및 하위 컴포넌트 수정]

wjdgotn77 committed:
[Refactor: Login.scss 코드리뷰 후 수정사항 반영]

wjdgotn77 committed:
- Login.js

 id와 password 의 state를 나눠 관리했던 것을 하나의 state에서 관리하는 것으로 수정하였습니다.
 id의 유효성 검사 로직을 더 직관적인 코드로 수정하였습니다.

wjdgotn77 committed:
[Refactor: Login.js 코드 리팩토링]

- Login.js
 1. id와 password 의 state를 나눠 관리했던 것을 하나의 state에서 관리하는 것으로 수정하였습니다.
 2. id의 유효성 검사 로직을 더 직관적인 코드로 수정하였습니다.
 3. BASE_URL을 추가하였습니다.

wjdgotn77 committed:
[Refactor: 코드리뷰 후 수정사항 반영]

wjdgotn77 committed:
[Refactor: 코드리뷰 후 리팩토링 진행중]

wjdgotn77 committed:
[Refactor: 코드리뷰 반영 class명 수정 및 불필요한 div 태그 정리]

wjdgotn77 committed:
[Refactor: 코드리뷰 반영 scss 수정]
